### PR TITLE
Add cr-ctc loss and ctc-decode in aishell

### PR DESCRIPTION
**Pull request by hhzzff**
1. Support CR-CTC loss in aishell.
2. Add ctc-decode for aishell

### CODE REVIEW

**Review by csukuangfj — APPROVED** [review]
Thank you for your contribution!

**Comment by csukuangfj** [comment]
Please fix the comments from the copilot.

File: egs/aishell/ASR/zipformer/ctc_decode.py [hunk 1]
```diff
@@ -0,0 +1,711 @@
...
+    )
+
+    parser.add_argument(
+        "--lang-dir",
+        type=Path,
+        default="data/lang_char",
+        help="The lang dir containing word table and LG graph",
+    )
+
+    parser.add_argument(
```
Comment (yaozengwei) [inline egs/aishell/ASR/zipformer/ctc_decode.py:156]: Could also support `ctc-prefix-beam-search` as in 
https://github.com/k2-fsa/icefall/blob/master/egs/librispeech/ASR/zipformer/ctc_decode.py#L486

File: egs/aishell/ASR/zipformer/ctc_decode.py [hunk 2]
```diff
@@ -0,0 +1,711 @@
...
+        "--nbest-scale",
+        type=float,
+        default=0.5,
+        help="""Scale applied to lattice scores when computing nbest paths.
+        Used only when the decoding method is and fast_beam_search_nbest_oracle""",
+    )
+
+    parser.add_argument(
```
Comment (yaozengwei) [inline egs/aishell/ASR/zipformer/ctc_decode.py:259]: It seems we don't use `blank_penalty` in ctc decoding.  Could remove this.

File: egs/aishell/ASR/zipformer/ctc_decode.py [hunk 3]
```diff
@@ -0,0 +1,711 @@
...
+import os
+from collections import defaultdict
+from pathlib import Path
+from typing import Dict, List, Optional, Tuple
+
+import k2
+import torch
+import torch.nn as nn
+from asr_datamodule import AishellAsrDataModule
+from beam_search import (
```
Comment (yaozengwei) [inline egs/aishell/ASR/zipformer/ctc_decode.py:50]: These methods are for transducer models. Could remove them.

File: egs/aishell/ASR/zipformer/ctc_decode.py [hunk 4]
```diff
@@ -0,0 +1,711 @@
...
+        default="ctc-greedy-search",
+        help="""Decoding method.
+        Supported values are:
+        - (1) ctc-greedy-search. Use CTC greedy search. It uses a sentence piece
+          model, i.e., lang_dir/bpe.model, to convert word pieces to words.
+          It needs neither a lexicon nor an n-gram LM.
+        """,
+    )
+
+    parser.add_argument(
```
Comment (yaozengwei) [inline egs/aishell/ASR/zipformer/ctc_decode.py:168]: Please check the unused arguments and do a clean up.

File: egs/aishell/ASR/zipformer/ctc_decode.py [hunk 5]
```diff
@@ -0,0 +1,711 @@
...
+                    filename_start=filename_start,
+                    filename_end=filename_end,
+                    device=device,
+                )
+            )
+
+    model.to(device)
+    model.eval()
+
+    if "fast_beam_search" in params.decoding_method:
```
Comment (yaozengwei) [inline egs/aishell/ASR/zipformer/ctc_decode.py:651]: Could remove.

**Comment by hhzzff** [comment]
experimenting feats related to ctc-prefix-beam-search, current code not for merge

File: egs/aishell/ASR/zipformer/ctc_decode.py [hunk 6]
```diff
@@ -0,0 +1,564 @@
...
+        Supported values are:
+        - (1) ctc-greedy-search. Use CTC greedy search. It uses a sentence piece
+          model, i.e., lang_dir/bpe.model, to convert word pieces to words.
+          It needs neither a lexicon nor an n-gram LM.
+          (2) ctc-prefix-beam-search. Extract n paths with the given beam, the best
+          path of the n paths is the decoding result.
+        """,
+    )
+
+    parser.add_argument(
```
Comment (yaozengwei) [inline egs/aishell/ASR/zipformer/ctc_decode.py:158]: This is not used.

File: egs/aishell/ASR/zipformer/ctc_decode.py [hunk 7]
```diff
@@ -0,0 +1,564 @@
...
+    add_model_arguments(parser)
+
+    return parser
+
+
+def get_decoding_params() -> AttributeDict:
+    """Parameters for decoding."""
+    params = AttributeDict(
+        {
+            "frame_shift_ms": 10,
```
Comment (yaozengwei) [inline egs/aishell/ASR/zipformer/ctc_decode.py:174]: These are not used. Could check again.

File: egs/aishell/ASR/zipformer/ctc_decode.py [hunk 8]
```diff
@@ -0,0 +1,564 @@
...
+    model.eval()
+
+    num_param = sum([p.numel() for p in model.parameters()])
+    logging.info(f"Number of model parameters: {num_param}")
+
+    # we need cut ids to display recognition results.
+    args.return_cuts = True
+    aishell = AishellAsrDataModule(args)
+
+    def remove_short_utt(c: Cut):
```
Comment (yaozengwei) [inline egs/aishell/ASR/zipformer/ctc_decode.py:527]: I am curious whether this filter works in testing. Did your testing log print "Exclude cut with ID** from decoding**". If it really did exclude somes test samples, would it raise concerns about unfair comparison to other models like Conformer? I notice `decode.py` also has this behavior.

Comment (hhzzff) [reply egs/aishell/ASR/zipformer/ctc_decode.py]: Logs show that the function didn't cut any short utturance in datasets. I also checked decoding results in https://huggingface.co/zrjin/icefall-asr-aishell-zipformer-2023-10-24, which didn't work as well. Latest commits have removed this function.

**Review by yaozengwei — APPROVED** [review]

File: egs/aishell/ASR/RESULTS.md [hunk 9]
```diff
@@ -941,3 +941,61 @@ The best decoding results (CER) are listed below, we got this results by averagi
 ||test|
 |--|--|
 |CER| 10.16% |
+
+### Aishell training results (zipformer + CR-CTC)
```
[the remark this answers is not in the document]
Comment (hhzzff) [reply egs/aishell/ASR/RESULTS.md]: Done.

[the remark this answers is not in the document]
Comment (yaozengwei) [reply egs/aishell/ASR/RESULTS.md]: Thanks!
